File: src/trading/evmSwap.js
// src/trading/evmSwap.js
const { ethers } = require('ethers');
const axios = require('axios');
const { getProvider } = require('../wallet/balanceChecker');
const { getEVMSigner } = require('../wallet/walletGenerator');
const { CHAINS } = require('../config/chains');
const logger = require('../utils/logger');

const UNISWAP_V2_ABI = [
  'function swapExactETHForTokens(uint amountOutMin, address[] calldata path, address to, uint deadline) external payable returns (uint[] memory amounts)',
  'function swapExactTokensForETH(uint amountIn, uint amountOutMin, address[] calldata path, address to, uint deadline) external returns (uint[] memory amounts)',
  'function swapExactTokensForTokens(uint amountIn, uint amountOutMin, address[] calldata path, address to, uint deadline) external returns (uint[] memory amounts)',
  'function getAmountsOut(uint amountIn, address[] calldata path) external view returns (uint[] memory amounts)'
];

const ERC20_ABI = [
  'function approve(address spender, uint256 amount) external returns (bool)',
  'function allowance(address owner, address spender) external view returns (uint256)',
  'function balanceOf(address) view returns (uint256)',
  'function decimals() view returns (uint8)',
  'function symbol() view returns (string)'
];

const NATIVE_TOKEN = '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE';

/**
 * Get a swap quote using 1inch aggregator API
 */
async function get1inchQuote(chainId, tokenIn, tokenOut, amountIn, slippage = 1) {
  if (!process.env.ONEINCH_API_KEY || process.env.ONEINCH_API_KEY.includes('your_')) {
    return null;
  }

  try {
    const res = await axios.get(`https://api.1inch.dev/swap/v6.0/${chainId}/quote`, {
      params: { src: tokenIn, dst: tokenOut, amount: amountIn.toString() },
      headers: { Authorization: `Bearer ${process.env.ONEINCH_API_KEY}` },
      timeout: 8000
    });
    return res.data;
  } catch (err) {
    logger.warn(`1inch quote failed: ${err.message}`);
    return null;
  }
}

/**
 * Get a swap transaction from 1inch
 */
async function get1inchSwapTx(chainId, tokenIn, tokenOut, amountIn, fromAddress, slippage = 1) {
  if (!process.env.ONEINCH_API_KEY || process.env.ONEINCH_API_KEY.includes('your_')) {
    return null;
  }

  try {
    const res = await axios.get(`https://api.1inch.dev/swap/v6.0/${chainId}/swap`, {
      params: {
        src: tokenIn, dst: tokenOut,
        amount: amountIn.toString(),
        from: fromAddress,
        slippage: slippage.toString(),
        disableEstimate: false
      },
      headers: { Authorization: `Bearer ${process.env.ONEINCH_API_KEY}` },
      timeout: 10000
    });
    return res.data.tx;
  } catch (err) {
    logger.warn(`1inch swap tx failed: ${err.message}`);
    return null;
  }
}

/**
 * Get quote from UniswapV2-compatible DEX
 */
async function getV2Quote(chainKey, tokenIn, tokenOut, amountIn) {
  const chain = CHAINS[chainKey];
  const provider = getProvider(chainKey);
  const router = new ethers.Contract(chain.router, UNISWAP_V2_ABI, provider);

  const isNativeIn = tokenIn.toLowerCase() === NATIVE_TOKEN.toLowerCase();
  const isNativeOut = tokenOut.toLowerCase() === NATIVE_TOKEN.toLowerCase();

  const actualIn = isNativeIn ? chain.wrappedNative : tokenIn;
  const actualOut = isNativeOut ? chain.wrappedNative : tokenOut;

  const path = [actualIn, actualOut];
  const amounts = await router.getAmountsOut(amountIn, path);
  return amounts[amounts.length - 1];
}

/**
 * Ensure token approval for router
 */
async function ensureApproval(signer, tokenAddress, spender, amount) {
  const token = new ethers.Contract(tokenAddress, ERC20_ABI, signer);
  const allowance = await token.allowance(await signer.getAddress(), spender);

  if (allowance < amount) {
    logger.info(`Approving ${tokenAddress} for ${spender}`);
    const approveTx = await token.approve(spender, ethers.MaxUint256);
    await approveTx.wait();
    logger.info(`Approval confirmed: ${approveTx.hash}`);
  }
}

/**
 * Execute swap on EVM chain
 * Returns { txHash, amountOut, status }
 */
async function executeEVMSwap({
  chainKey,
  encryptedPrivateKey,
  tokenIn,
  tokenOut,
  amountIn,        // BigInt or string in wei
  slippagePercent = 1.0,
  userAddress
}) {
  const chain = CHAINS[chainKey];
  if (!chain || chain.type !== 'evm') throw new Error(`Invalid EVM chain: ${chainKey}`);

  const provider = getProvider(chainKey);
  const signer = getEVMSigner(encryptedPrivateKey, provider);
  const fromAddress = await signer.getAddress();

  const amountInBN = BigInt(amountIn.toString());
  const isNativeIn = tokenIn.toLowerCase() === NATIVE_TOKEN.toLowerCase();
  const isNativeOut = tokenOut.toLowerCase() === NATIVE_TOKEN.toLowerCase();

  logger.info(`Executing swap on ${chainKey}: ${tokenIn} → ${tokenOut}, amount: ${amountIn}`);

  // Try 1inch first if API key is configured
  const oneinchTx = await get1inchSwapTx(
    chain.id, tokenIn, tokenOut, amountInBN.toString(), fromAddress, slippagePercent
  );

  if (oneinchTx) {
    logger.info('Using 1inch aggregator for swap');
    const tx = await signer.sendTransaction({
      to: oneinchTx.to,
      data: oneinchTx.data,
      value: BigInt(oneinchTx.value || '0'),
      gasLimit: BigInt(Math.floor(Number(oneinchTx.gas) * 1.2))
    });
    const receipt = await tx.wait();
    return { txHash: tx.hash, status: receipt.status === 1 ? 'success' : 'failed' };
  }

  // Fallback to direct DEX router
  logger.info('Using direct DEX router for swap');
  const router = new ethers.Contract(chain.router, UNISWAP_V2_ABI, signer);
  const deadline = Math.floor(Date.now() / 1000) + 1200; // 20 min

  const actualIn = isNativeIn ? chain.wrappedNative : tokenIn;
  const actualOut = isNativeOut ? chain.wrappedNative : tokenOut;
  const path = [actualIn, actualOut];

  // Get expected output with slippage
  const amountsOut = await router.getAmountsOut(amountInBN, path);
  const expectedOut = amountsOut[amountsOut.length - 1];
  const minOut = expectedOut * BigInt(Math.floor((100 - slippagePercent) * 100)) / 10000n;

  let tx;

  if (isNativeIn) {
    // ETH → Token
    tx = await router.swapExactETHForTokens(minOut, path, fromAddress, deadline, {
      value: amountInBN,
      gasLimit: 300000n
    });
  } else if (isNativeOut) {
    // Token → ETH
    await ensureApproval(signer, tokenIn, chain.router, amountInBN);
    tx = await router.swapExactTokensForETH(amountInBN, minOut, path, fromAddress, deadline, {
      gasLimit: 300000n
    });
  } else {
    // Token → Token
    await ensureApproval(signer, tokenIn, chain.router, amountInBN);
    tx = await router.swapExactTokensForTokens(amountInBN, minOut, path, fromAddress, deadline, {
      gasLimit: 400000n
    });
  }

  logger.info(`Swap tx submitted: ${tx.hash}`);
  const receipt = await tx.wait();

  return {
    txHash: tx.hash,
    amountOut: expectedOut.toString(),
    status: receipt.status === 1 ? 'success' : 'failed'
  };
}

/**
 * Send native token (ETH/BNB/MATIC etc)
 */
async function sendNativeToken(chainKey, encryptedPrivateKey, toAddress, amountEther) {
  const provider = getProvider(chainKey);
  const signer = getEVMSigner(encryptedPrivateKey, provider);

  const tx = await signer.sendTransaction({
    to: toAddress,
    value: ethers.parseEther(amountEther.toString())
  });

  const receipt = await tx.wait();
  return { txHash: tx.hash, status: receipt.status === 1 ? 'success' : 'failed' };
}

/**
 * Send ERC20 token
 */
async function sendERC20Token(chainKey, encryptedPrivateKey, tokenAddress, toAddress, amount, decimals) {
  const provider = getProvider(chainKey);
  const signer = getEVMSigner(encryptedPrivateKey, provider);
  const token = new ethers.Contract(tokenAddress, ERC20_ABI, signer);

  const amountRaw = ethers.parseUnits(amount.toString(), decimals);
  const tx = await token.transfer(toAddress, amountRaw);
  const receipt = await tx.wait();

  return { txHash: tx.hash, status: receipt.status === 1 ? 'success' : 'failed' };
}

module.exports = {
  executeEVMSwap,
  sendNativeToken,
  sendERC20Token,
  get1inchQuote,
  getV2Quote
};

File: python_bot/trading/evm_swap.py
# trading/evm_swap.py  —  Live EVM swaps via Uniswap V2 / 1inch

import json
from web3 import Web3
from wallet.balances import get_w3, ERC20_ABI
from wallet.generator import get_evm_signer
from config.chains import CHAINS
from config.secrets import ONEINCH_API_KEY
import requests

NATIVE = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"

UNIV2_ROUTER_ABI = [
    {"name": "swapExactETHForTokens", "type": "function", "stateMutability": "payable",
     "inputs": [{"name": "amountOutMin","type":"uint256"},{"name":"path","type":"address[]"},
                {"name":"to","type":"address"},{"name":"deadline","type":"uint256"}],
     "outputs": [{"type": "uint256[]"}]},
    {"name": "swapExactTokensForETH", "type": "function", "stateMutability": "nonpayable",
     "inputs": [{"name":"amountIn","type":"uint256"},{"name":"amountOutMin","type":"uint256"},
                {"name":"path","type":"address[]"},{"name":"to","type":"address"},
                {"name":"deadline","type":"uint256"}],
     "outputs": [{"type": "uint256[]"}]},
    {"name": "swapExactTokensForTokens","type":"function","stateMutability":"nonpayable",
     "inputs": [{"name":"amountIn","type":"uint256"},{"name":"amountOutMin","type":"uint256"},
                {"name":"path","type":"address[]"},{"name":"to","type":"address"},
                {"name":"deadline","type":"uint256"}],
     "outputs": [{"type": "uint256[]"}]},
    {"name": "getAmountsOut","type":"function","stateMutability":"view",
     "inputs": [{"name":"amountIn","type":"uint256"},{"name":"path","type":"address[]"}],
     "outputs": [{"type": "uint256[]"}]},
]

def _get_1inch_swap(chain_id, token_in, token_out, amount, from_addr, slippage=1):
    if not ONEINCH_API_KEY:
        return None
    try:
        r = requests.get(
            f"https://api.1inch.dev/swap/v6.0/{chain_id}/swap",
            params={"src": token_in, "dst": token_out, "amount": str(amount),
                    "from": from_addr, "slippage": str(slippage)},
            headers={"Authorization": f"Bearer {ONEINCH_API_KEY}"},
            timeout=10
        )
        if r.status_code == 200:
            return r.json().get("tx")
    except Exception:
        pass
    return None

def _ensure_approval(w3, signer, token_addr, spender, amount):
    contract = w3.eth.contract(address=Web3.to_checksum_address(token_addr), abi=ERC20_ABI)
    allowance = contract.functions.allowance(signer.address, spender).call()
    if allowance < amount:
        tx = contract.functions.approve(spender, 2**256 - 1).build_transaction({
            "from": signer.address,
            "nonce": w3.eth.get_transaction_count(signer.address),
            "gasPrice": w3.eth.gas_price,
        })
        signed = signer.sign_transaction(tx)
        txh = w3.eth.send_raw_transaction(signed.raw_transaction)
        w3.eth.wait_for_transaction_receipt(txh)

def execute_evm_swap(chain_key: str, enc_key: str, token_in: str,
                     token_out: str, amount_wei: int, slippage: float = 1.0) -> dict:
    """Execute a real EVM swap. Returns {tx_hash, status}."""
    chain = CHAINS[chain_key]
    w3 = get_w3(chain_key)
    signer = get_evm_signer(enc_key)
    is_native_in = token_in.lower() == NATIVE.lower()
    is_native_out = token_out.lower() == NATIVE.lower()

    # Try 1inch first
    oneinch_tx = _get_1inch_swap(chain["id"], token_in, token_out, amount_wei, signer.address, slippage)
    if oneinch_tx:
        tx = {
            "to": oneinch_tx["to"],
            "data": oneinch_tx["data"],
            "value": int(oneinch_tx.get("value", 0)),
            "gas": int(oneinch_tx.get("gas", 300000) * 1.2),
            "gasPrice": w3.eth.gas_price,
            "nonce": w3.eth.get_transaction_count(signer.address),
            "chainId": chain["id"],
        }
        signed = signer.sign_transaction(tx)
        txh = w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = w3.eth.wait_for_transaction_receipt(txh, timeout=120)
        return {"tx_hash": txh.hex(), "status": "success" if receipt.status == 1 else "failed"}

    # Fallback: direct Uniswap V2 router
    router_addr = Web3.to_checksum_address(chain["router"])
    router = w3.eth.contract(address=router_addr, abi=UNIV2_ROUTER_ABI)
    deadline = w3.eth.get_block("latest")["timestamp"] + 1200
    actual_in = chain["wrapped"] if is_native_in else token_in
    actual_out = chain["wrapped"] if is_native_out else token_out
    path = [Web3.to_checksum_address(actual_in), Web3.to_checksum_address(actual_out)]

    amounts_out = router.functions.getAmountsOut(amount_wei, path).call()
    expected = amounts_out[-1]
    min_out = int(expected * (100 - slippage) / 100)

    nonce = w3.eth.get_transaction_count(signer.address)
    gas_price = w3.eth.gas_price

    if is_native_in:
        tx = router.functions.swapExactETHForTokens(
            min_out, path, signer.address, deadline
        ).build_transaction({"from": signer.address, "value": amount_wei,
                             "gas": 300000, "gasPrice": gas_price, "nonce": nonce, "chainId": chain["id"]})
    elif is_native_out:
        _ensure_approval(w3, signer, token_in, router_addr, amount_wei)
        tx = router.functions.swapExactTokensForETH(
            amount_wei, min_out, path, signer.address, deadline
        ).build_transaction({"from": signer.address, "value": 0,
                             "gas": 300000, "gasPrice": gas_price, "nonce": nonce, "chainId": chain["id"]})
    else:
        _ensure_approval(w3, signer, token_in, router_addr, amount_wei)
        tx = router.functions.swapExactTokensForTokens(
            amount_wei, min_out, path, signer.address, deadline
        ).build_transaction({"from": signer.address, "value": 0,
                             "gas": 400000, "gasPrice": gas_price, "nonce": nonce, "chainId": chain["id"]})

    signed = signer.sign_transaction(tx)
    txh = w3.eth.send_raw_transaction(signed.raw_transaction)
    receipt = w3.eth.wait_for_transaction_receipt(txh, timeout=120)
    return {"tx_hash": txh.hex(), "status": "success" if receipt.status == 1 else "failed",
            "amount_out": str(expected)}

def send_native(chain_key: str, enc_key: str, to: str, amount_ether: float) -> dict:
    w3 = get_w3(chain_key)
    signer = get_evm_signer(enc_key)
    amount_wei = Web3.to_wei(amount_ether, "ether")
    tx = {"to": Web3.to_checksum_address(to), "value": amount_wei,
          "gas": 21000, "gasPrice": w3.eth.gas_price,
          "nonce": w3.eth.get_transaction_count(signer.address),
          "chainId": CHAINS[chain_key]["id"]}
    signed = signer.sign_transaction(tx)
    txh = w3.eth.send_raw_transaction(signed.raw_transaction)
    receipt = w3.eth.wait_for_transaction_receipt(txh, timeout=120)
    return {"tx_hash": txh.hex(), "status": "success" if receipt.status == 1 else "failed"}
